### PR TITLE
Issue 5880: Reduce threads spawned while creating a StateSynchronizer and ReaderGroupManager.

### LINKED ISSUE #5880: Reduce the resources used by ReaderGroupManager and StateSyncrhonizer.
**Describe the bug**
Reduce threads spawned while creating a StateSynchronizer and ReaderGroupManager...

**To Reproduce**
Ensure the defaults used by StateSyncrhonizer and RederGroupManager are toned down.

**Screenshots**
NA

**Additional information**
NA

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
Reduce threads spawned while creating a StateSynchronizer and ReaderGroupManager.

**Purpose of the change**  
Fixes #5880 .

**What the code does**  
Reduce threads spawned while creating a StateSynchronizer and ReaderGroupManager.
- Configure connection pooling to use a max of 1 connection per segment store.
- Reduce the number of internal thread pools to 1.

**How to verify it**  
All the existing tests should continue to pass.

Note: Verification is in progress.

### CODE REVIEW

**Comment by shrids** [comment]
System test run pravega-branch-builder/1038 is in progress.

File: client/src/main/java/io/pravega/client/SynchronizerClientFactory.java [hunk 1]
```diff
@@ -36,9 +36,15 @@
      * @return Instance of ClientFactory implementation.
      */
     static SynchronizerClientFactory withScope(String scope, ClientConfig config) {
-        val connectionFactory = new SocketConnectionFactoryImpl(config);
-        return new ClientFactoryImpl(scope, new ControllerImpl(ControllerImplConfig.builder().clientConfig(config).build(),
-                connectionFactory.getInternalExecutor()), connectionFactory);
+        // Change the max number of number of allowed connections to the segment store to 1.
+        val updatedConfig = config.toBuilder()
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/SynchronizerClientFactory.java:40]: I don't think this change is appropriate. Here a user has a setting in the config that they can set, and we are just overriding it to be 1. Given that a factory can be used to create many objects there ought to be some way to configure the number of connections. This removes that.

I think a better solution would be to change the ClientConfig object to track if the value set is the default. Then if it has not been set then we can make the default different for different paths, such as by setting 1 here.

File: client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java [hunk 2]
```diff
@@ -42,7 +44,13 @@ public static ReaderGroupManager withScope(String scope, URI controllerUri) {
      * @return Instance of Stream Manager implementation.
      */
     public static ReaderGroupManager withScope(String scope, ClientConfig clientConfig) {
-        return new ReaderGroupManagerImpl(scope, clientConfig, new SocketConnectionFactoryImpl(clientConfig));
+        // Change the max number of number of allowed connections to the segment store to 1.
+        val updatedClientConfig = clientConfig.toBuilder()
+                .maxConnectionsPerSegmentStore(1)
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java:49]: Same here.

**Review by tkaitchuck — CHANGES_REQUESTED** [review]

File: client/src/main/java/io/pravega/client/SynchronizerClientFactory.java (hunk 1 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/SynchronizerClientFactory.java:46]: Thanks, modified the code to ensure `ClientConfig `object is modified only if the default value is used.

File: client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java (hunk 2 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java:55]: fixed it.

**Review by tkaitchuck — APPROVED** [review]

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 3]
```diff
@@ -187,6 +193,7 @@ public boolean isEnableTlsToSegmentStore() {
         private boolean validateHostName = true;
 
         private boolean deriveTlsEnabledFromControllerURI = true;
+        private boolean canOverrideMaxConnectionsConfiguration = true;
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/ClientConfig.java]: Thanks I have improved the code...

**Comment by shrids** [comment]
test-pravega-continuous-1/1395/ is in progress.

**Pull request merged**